drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2023 EPAM Systems
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT regulator_gpio
+
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(regulator_gpio, CONFIG_REGULATOR_LOG_LEVEL);
+
+struct regulator_gpio_config {
+	struct regulator_common_config common;
+	struct gpio_dt_spec switch_volt;
+	int initial_state;
+	const int32_t *states;
+	const uint8_t states_cnt;
+};
+
+struct regulator_gpio_data {
+	struct regulator_common_data common;
+	int32_t current_volt_uv;
+};
+
+static unsigned int regulator_gpio_count_voltages(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+
+	return cfg->states_cnt / 2;
```

Comment:
`2U`

Reply:
What of the purpose of this change `2U`?

> suggestion: why don't you store states / 2 to states_cnt? you seem to be dividing by 2 in more than one place
will fix, thanks

Reply:
done